crates/chainql-core/src/lib.rs

```diff
@@ -430,6 +432,69 @@ where
 	Ok(())
 }
 
+trait DecodeExt
+where
+	Self: parity_scale_codec::Decode,
+{
+	fn decode<I: Input>(input: &mut I) -> Result<Self> {
+		let remaining = input
+			.remaining_len()
+			.map_err(|err| runtime_error!("codec: failed to read remaining len: {err}"))?;
+
+		match <Self as parity_scale_codec::Decode>::decode(input) {
+			Ok(value) => Ok(value),
+			Err(err) => {
+				if err == "Not enough data to fill buffer".into()
+					&& let Some(expected) = Self::encoded_fixed_size()
+					&& let Some(remaining) = remaining
+				{
+					Err(runtime_error!(
+						"codec: not enough bytes to decode {}: expected {} bytes, but only {} byte is available",
+						core::any::type_name::<Self>(),
+						expected,
+						remaining
+					))
+				} else {
+					Err(runtime_error!("codec: {err}"))
+				}
+			}
+		}
+	}
+}
+
+impl<T> DecodeExt for T where T: parity_scale_codec::Decode {}
+
+/// Decode some value [`dec`] into the type [`typ`] in the registry [`reg`].
+/// Returns default if decode failed.
+fn decode_value_or_default<I>(
+	dec: &mut I,
+	reg: &PortableRegistry,
+	name: &str,
+	typ: UntrackedSymbol<TypeId>,
+	compact: bool,
+	default: Option<Vec<u8>>,
+	use_default_for_corrupted_storages: bool,
+) -> Result<Val>
+where
+	I: Input,
+{
+	let mut value = decode_value(dec, reg, typ, compact);
+
+	if let Err(err) = &value
+		&& use_default_for_corrupted_storages
+	{
+		warn!("Storage {name} is corrupted. Default values will be used. Error: {err}");
```

Comment:
В tracing все сообщения пишут с маленькой буквы.
```suggestion
		warn!("storage {name} is corrupted, default values will be used: {err}");
```

Использование ToString для jsonnetовых ошибок это не очень хорошо, и правильнее было бы сюда пробросить TraceFormat... Но слишком много кода, пусть остаётся, оно не должно так часто вызываться...

Reply:
Я помню, что ToString не очень хорошо использовать для ошибок jrsonnet. Но TraceFormat настолько неприятно прокидывать было, что оставил вот так. В принципе из сообщения будет примерно понятно, что пошло не по плану